### PR TITLE
FP: 2716 + 2711 + 2927

### PR DESCRIPTION
**Pull request by quirinpa**
- [FP-2716](https://movai.atlassian.net/browse/FP-2716): Annotations imported cannot be edited
- [FP-2711](https://movai.atlassian.net/browse/FP-2711): Corrupted data in Annotation
- [FP-2927](https://movai.atlassian.net/browse/FP-2927): IDE - Validate user input data string/array/object and only try to transform when needed
- [FP-2934](https://movai.atlassian.net/browse/FP-2934): Cursor position resets when editing a flow parameter (possibly but not officially)

Needs: https://github.com/MOV-AI/frontend-npm-ide/pull/605

[FP-2716]: https://movai.atlassian.net/browse/FP-2716?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[FP-2711]: https://movai.atlassian.net/browse/FP-2711?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[FP-2927]: https://movai.atlassian.net/browse/FP-2927?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[FP-2934]: https://movai.atlassian.net/browse/FP-2934?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by diasdm — CHANGES_REQUESTED** [review]
Creating an annotation results in an escaped array
image

**Comment by quirinpa** [comment]
[That](https://github.com/MOV-AI/frontend-npm-lib-ide/pull/338#pullrequestreview-2254519001) is fixed there: https://github.com/MOV-AI/frontend-npm-ide/pull/605

**Review by diasdm — CHANGES_REQUESTED** [review]
There's an exception when opening an annotation with escaped array:
image

**Review by manuelnogueiramov — DISMISSED** [review]
Did a code review, not test: Just 2 small suggestions that won't hold this PR. Good work @quirinpa :tada:

File: src/editors/_shared/hooks/DataTypes/AbstractDataType.js [hunk 1]
```diff
@@ -118,13 +118,21 @@ class AbstractDataType {
...
-        value={value || ""}
-        onChange={evt => props.onChange(evt.target.value)}
+        defaultValue={(convert ? JSON.stringify(value) : value) || ""}
+        onChange={evt => {
+          if (!convert)
+            return props.onChange(evt.target.value);
+
+          try {
+            props.onChange(JSON.parse(evt.target.value));
+          } catch (e) {}
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/AbstractDataType.js:134]: Maybe we could move the above onChange call to it. You know, this still needs a bit of work, since I could make this work without the issues, but its still not ideal in the way that it works. Because it defaults to string type values, instead of the original type..

I'll take this into account when finishing it up

[the remark this answers is not in the document]
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/AbstractDataType.js]: That's resolved now

**Review by diasdm — CHANGES_REQUESTED** [review]
The library is not building

File: src/editors/_shared/hooks/DataTypes/types/StringType.js [hunk 2]
```diff
@@ -23,20 +22,8 @@ class StringType extends DataType {
...
-        const parsed = this.getParsedValue(value);
-        const isValid =
-          typeof parsed === DATA_TYPES.STRING || parsed instanceof String;
-        resolve({ success: isValid });
-      } catch (e) {
-        resolve({ success: false, error: "MandatoryStringQuotes" });
-      }
-    });
+  validate(_value) {
+    return Promise.resolve({ success: true });
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/types/StringType.js:26]: I don't like that we aren't validating this anymore. there were a lot of validations taking place. Removing these validations might end up causing regressions. Can you please confirm why this was there in the first place and if it is safe to remove?

File: src/editors/_shared/hooks/DataTypes/types/ArrayType.js [hunk 3]
```diff
@@ -22,7 +22,7 @@ class ArrayType extends DataType {
    * @returns
    */
   validate(value) {
-    value = value?.replace(/'/g, '"');
+    value = typeof value === "string" ? value?.replace(/'/g, '"'): value;
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/types/ArrayType.js:25]:
```suggestion
    value = typeof value === "string" ? value.replace(/'/g, '"'): value;
```
Don't need the `?` in `value.replace` because we already know it's of `string` type, so it won't be `null` / `undefined`.

**Review by manuelnogueiramov — CHANGES_REQUESTED** [review]
1 Small suggestion, and 1 question that I need answering please.

File: src/editors/_shared/hooks/DataTypes/types/StringType.js (hunk 2 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/types/StringType.js:26]: It's really doing nothing. Since strings are always strings now.

**Review by diasdm — COMMENTED** [review]
Functionally wise, only thing not working as expected is that when saving an old string object, the value is not saved as array

File: src/editors/_shared/hooks/DataTypes/AbstractDataType.js [hunk 4]
```diff
@@ -117,14 +117,23 @@ class AbstractDataType {
    * @returns {ReactComponent} Text input for editing common strings
    */
   stringEditComponent(props, placeholder, parsedValue) {
```
Comment (diasdm) [inline src/editors/_shared/hooks/DataTypes/AbstractDataType.js:119]: It seems to me this function is trying to do too much with too little information. Maybe creating a arrayEditComponent and objectEditComponent would be more appropriate? Or parametrize the function to save the values with the expected type?

**Comment by quirinpa** [comment]
Added the other task and now it is all fixed

**Review by diasdm — CHANGES_REQUESTED** [review]
When adding an entry and selecting type string, if in the value we input `5454` or `["a": 1]` the value saved is an int and array respectively. It should be a string

**Comment by quirinpa** [comment]
That doesn't happen as you forgot to pull. No JSON.parse happens for strings now

**Comment by quirinpa** [comment]
That's natural. We should not hold the PR because of it.
It should be fixed manually, if we're editing it.

**Review by diasdm — CHANGES_REQUESTED** [review]
Two issues present:
1. Objects in annotation
image
2. Strings in annotation saved escaped
image

**Comment by quirinpa** [comment]
No quotes required.

**Comment by diasdm** [comment]
> No quotes required. And that's not an object. Its an invention. Nice invention David

I always wanted to be an inventor :D
Jokes aside, the test was writing an invalid object and I didn't realize, sorry
But quoting strings is mandatory with the current implementation required:
image

**Comment by manuelnogueiramov** [comment]
These first tests were all done in a node.

Strings:

image

I'm not sure if I did anything wrong, but running this locally I get this error when trying to add a string without quotes. 
If I then add the quotes, I get this:

image

testing with empty string (couldn't save as empty, had to add the "") got this: `"\"\""`.

Using Any was ok:

image

----------------------------------------
Numbers:

Attempted to save an empty number and was successful. But I don't think it should... 
image

Since it's a number, at least 0 should be saved, right? That's the default number and that's how it worked before, not sure if this is correct or not.

----------------------------------------

Bools: 

Saved a bool as false:
image

But when I click to edit I see it as true:
image

If I then don't do anything and try to save I get:
image

Just out of curiosity, tried to change from bool to another type (ie: object) and this happened:
image

And it closed the modal, and stayed as bool.

-----------------------------------------
Objects:

When I picked object this happened:

image

Creating an object like the one in the image fails the validation (which is the same as 1.2.2-1):

image

other than that worked fine.

-----------------------------------------
Arrays:

When I picked array this happened:

image

Other than that, arrays worked good:
image
image


-----------------------------------------
Configurations:

worked good.

-----------------------------------------

Now onto the flows:

~Just opening a parameter and then saving (without doing any changes), removes the default value from it. When you open it again, it selected Custom Value and has an empty value.~ -> This is indeed a bug that wasn't caused by this PR, opened a [ticket for it here](https://movai.atlassian.net/browse/FP-2948).

Trying to disable a string, number, boolean, array, object parameter got this:
image

Trying to dissable a configuration parameter got this:
image


-----------------------------------------

Now onto the annotations:

Trying to save a string without quotes is invalid. Saving it with quotes, adds escapes.

Trying to save a number with the default 0 is invalid, you have to specifically go up to 1 and then back to 0.
Trying to save a number like 000001 (just saves 1) or 000000 (just saves 0) doesn't work. In 1.2.2-1 it saves all annotations as strings. Not sure which one is the correct one, @diasdm, @diasnad?

Other than that everything else seemed to work nicely.

Sorry for the long test, but this is a sensitive PR and should be treated as such. Please make sure you're not breaking any functionalities as to not introduce more regressions. 
It's better to take more time but deliver something that you are proud of and know it's not breaking stuff, than to be the fastest guy delivering stuff, but then breaks other things. Take a look at annotations, nodes, flows, wherever these files are being used at and make sure their functionality is working as intended.
Make sure to annoy @diasdm and @diasnad to get the best outcome. Maybe some people from robotics? CSAI? that use the platform might be able to say what they really need from these functionalities. 
If there's no real clear path, just keep what we already had working and try to make smaller changes as to just fix what is needed.

P.S. I haven't reviewed the code past the second file, because I noticed a lot of changes and decided to test it first to check if any functionality was lost, and yeah. This is the result. Feel free to PM me if you have any questions or need a pair programming session where we can both try and fix this without causing regressions.

**Review by manuelnogueiramov — CHANGES_REQUESTED** [review]
A lot of functionality lost, regression bugs will come out of this. Please pay more attention to these sensitive PR's.

**Comment by quirinpa** [comment]
Yes but why require quotes if inputs already return strings. And what happens if you need quotes in the string?

**Comment by quirinpa** [comment]
For JSON objects to work, you need to quote the string keys just like before.
0 is the default for numbers. If there's no change, there's no need to save. You couldn't save the empty string because it is the default. Again. I can try to fix this by changing it so it saves the default explicitly and shows it.
I tried changing a bool to an object, and it didn't give out the screenshot you shown. How did you take it if it closed the modal?
I can see that the modal closes. And that that is undesirable. I can handle that by changing the value to the default in that situation.
I acknowledge there's issues with the Boolean type on nodes. Namely about the state of the checkbox not matching the value when it is false. However the value is saved correctly.
I never tried disabling a parameter. So that checks out.
The error boundary errors don't show up in production.

**Comment by quirinpa** [comment]
All fixed now. It's pretty natural that 0001 doesn't work as a number. Because that's not what it is.
The only thing I haven't tried is that about disabled values.

**Comment by diasdm** [comment]
> Not sure which one is the correct one, @diasdm, @diasnad?

Currently all parameters are saved as strings see [here](https://github.com/MOV-AI/qa_ui_tests/blob/main/movai_qa_ui_tests/metadata/Annotation/new_annotation.json), lets keep this behavior for the time being

**Comment by quirinpa** [comment]
All done. Please re-review

**Review by diasdm — CHANGES_REQUESTED** [review]
In the node editor, adding a string param results in a different output than before
image

In the annotations editor, adding a configuration param is not possible

https://github.com/user-attachments/assets/6ee28654-b0ab-4b3c-b37f-ee916446dd1e

**Comment by quirinpa** [comment]
The first isn't a bug. But the second is.

File: src/editors/_shared/hooks/DataTypes/testUtils.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  it("Set value correctly", () => {
+    const onChange = jest.fn();
+    getRendered(
+      new DataType(), {
+        rowData: { value: [] },
+        onChange,
+      }
+    );
+    const input = screen.getByTestId('input_value');
+    fireEvent.change(input, { target: { value: stringValue } });
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/testUtils.js:26]: `fireEvent` is not recommended. Please use `userEvent`. As mentioned [here](https://blog.mimacom.com/react-testing-library-fireevent-vs-userevent/)
TL;DR; 
**So, what's the difference between them? Well, the difference is that userEvent simulates complete interactions, and not only single events.**

File: src/editors/_shared/hooks/DataTypes/types/AnyType.test.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  true, true, true, true, true, true, true,
+  true, true, true, true, true, true,
+  true, true, true, true, true, true, true,
+  true, true, true, true, true, true,
+  // onlyStrings
+  true, true, true, true, true, true, true,
+  true, true, true, true, true, true,
+  true, true, true, true, true, true, true,
+  true, true, true, true, true, true, true
+], "[1]", "[1]");
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/types/AnyType.test.js:14]: I think I understand what you are doing here, but this is a bit unbearable to actually know what to see when some test fails...
Maybe you can make the validation suit return an object with a small description of the test like `testValidObject1`, `testValidString1`, `testInvalidString1`, stuff like this. This way we know, when something fails, where it failed and why. Instead of having to compare an array of returns that we don't know by heart...

File: src/editors/_shared/hooks/DataTypes/types/ArrayType.js [hunk 7]
```diff
@@ -3,37 +3,12 @@ import DataType from "../AbstractDataType";
...
-   * @param {*} value
-   * @returns
-   */
-  validate(value) {
-    value = value?.replace(/'/g, '"');
-    return new Promise(resolve => {
-      try {
-        if (checkIfDefaultOrDisabled(value)) {
-          return resolve({ success: true, value });
-        }
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/types/ArrayType.js:30]: The bug that I mention in my final comment, might be avoided with this check here.

File: src/editors/_shared/hooks/DataTypes/types/BooleanType.js [hunk 8]
```diff
@@ -5,90 +5,45 @@ import { pythonToBool, boolToPython } from "../../../../../utils/Utils";
...
-    return isPythonValue ? boolToPython(false) : false;
+  parse(value) {
+    if (value === "")
+      return undefined;
+    if (value === "True")
+      return true;
+    if (value === "False")
+      return false;
+    return null;
   }
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/types/BooleanType.js:28]: In which cases we want to return `undefined` and in which ones we want to return `null`?
With this suggestion it'll always return `true` / `false` / `undefined` accordingly.

```suggestion
  parse(value) {
    const possibleValues = {
      "True": true,
      "False": false
    }
    return possibleValues[value];
  }
```

File: src/editors/_shared/hooks/DataTypes/types/NumberType.js [hunk 9]
```diff
@@ -5,60 +5,19 @@ import DataType from "../AbstractDataType";
...
-        const isValid = typeof parsed === DATA_TYPES.NUMBER && !isNaN(parsed);
-        resolve({ success: isValid, parsed });
-      } catch (e) {
-        resolve({ success: false });
-      }
-    });
+  _validate(value) {
+    return value === undefined
+      || value !== null && typeof value === this.key
+      && !isNaN(value) && !Array.isArray(value);
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/types/NumberType.js:16]: Mixing operators without a clear goal is always dangerous. 
Is this what you want?

```suggestion
    return value === undefined || 
      (value !== null && typeof value === this.key
      && !isNaN(value) && !Array.isArray(value));
```

File: src/editors/_shared/hooks/DataTypes/types/ObjectType.js [hunk 10]
```diff
@@ -3,36 +3,12 @@ import DataType from "../AbstractDataType";
...
-          return resolve({ success: true, value });
-        }
-        const parsed = JSON.parse(value);
-        resolve({ success: parsed.constructor === Object });
-      } catch (e) {
-        resolve({ success: false });
-      }
-    });
+  _validate(value) {
+    return value === undefined || value !== null && typeof value === "object" && !Array.isArray(value);
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/types/ObjectType.js:11]: Same thing here with the mixing of operators...

```suggestion
    return value === undefined || (value !== null && typeof value === "object" && !Array.isArray(value));
```

File: src/editors/_shared/hooks/useDataTypes.js [hunk 11]
```diff
@@ -1,78 +1,40 @@
...
   const theme = useTheme();
-  const dataTypeManager = new DataTypeManager({ theme });
-
-  //========================================================================================
-  /*                                                                                      *
-   *                                    Exposed Methods                                   *
-   *                                                                                      */
-  //========================================================================================
+  const dataTypeManager = useMemo(
+    () => new DataTypeManager({ theme, onlyStrings }),
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/useDataTypes.js:11]: I think you didn't do this. But why the hell we need to pass the `theme` to a class constructor?

File: src/utils/Utils.js [hunk 12]
```diff
@@ -117,43 +117,6 @@ export function validateDocumentName(name = "") {
   }
 }
 
-const boolToPythonOptions = {
-  true: "True",
-  false: "False"
-};
-
-const PythonToBoolOptions = {
-  True: true,
```
Comment (manuelnogueiramov) [inline src/utils/Utils.js:126]: You could also use these consts to help with validations in the boolean datatype.

**Review by manuelnogueiramov — CHANGES_REQUESTED** [review]
Kudos for the tests and for the .stories correction! Awesome @quirinpa :+1: 
That's the way we should move forward!

Now, I didn't quite like the tests because they aren't human readable. I mean if they start failing we have to try and decode an array and see which one failed, and... yeah, not feasible. Maybe turn the battery of tests into an object with some minor naming as descriptions. 

There are still some corner case issues:
image
If you create a parameter (Object, Array, String anything with the codeEditor) and give it the value of "None" it will get stuck there, as you can't change it's value anymore...

File: src/utils/Utils.js (hunk 12 above)
Comment (quirinpa) [reply src/utils/Utils.js:126]: No longer needed. We have parse and uparse for that

File: src/editors/_shared/hooks/useDataTypes.js (hunk 11 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/useDataTypes.js:11]: Because it goes to some of the edit components

File: src/editors/_shared/hooks/DataTypes/types/ObjectType.js (hunk 10 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/types/ObjectType.js:11]: There is priority to && over ||, but if you wish for () to make it clearer, I'm not against.

File: src/editors/_shared/hooks/DataTypes/types/NumberType.js (hunk 9 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/types/NumberType.js:16]: I've replied below

File: src/editors/_shared/hooks/DataTypes/types/BooleanType.js (hunk 8 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/types/BooleanType.js:28]: That is not equivalent.

File: src/editors/_shared/hooks/DataTypes/DataTypeManager.js [hunk 13]
```diff
@@ -7,7 +7,8 @@ import ObjectType from "./types/ObjectType";
 import StringType from "./types/StringType";
 
 class DataTypeManager {
-  constructor({ theme, selfProvision = true }) {
+  constructor({ theme, selfProvision = true, onlyStrings = false } = {}) {
+    this.onlyStrings = onlyStrings;
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/DataTypeManager.js:11]: What do you mean? Please clarify

File: src/editors/_shared/hooks/DataTypes/AbstractDataType.js [hunk 14]
```diff
@@ -1,25 +1,122 @@
...
+
+    const parsing = {
+      parse: this.parse.bind(this),
+      unparse: this.unparse.bind(this),
+    }, noParsing = { parse: identity, unparse: identity };
+
+    const doInputParsing = (onlyStrings ? !textInput : textInput);
+    this.parsing = doInputParsing ? parsing : noParsing;
+    this._validationParse = onlyStrings ? parsing.parse : identity;
+    this.getSaveable = onlyStrings ? parsing.unparse : identity;
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/AbstractDataType.js:92]: We need this.parsing like it is. Check the edit components

File: src/editors/_shared/hooks/DataTypes/AbstractDataType.js [hunk 15]
```diff
@@ -1,25 +1,122 @@
...
+  constructor({ theme, onlyStrings, textInput = true } = {}) {
     // Set hooks to be used in abstract renders
     this._theme = theme;
-    this.defaultStringEditor = this.defaultStringEditor.bind(this);
+    this._onlyStrings = onlyStrings;
+
+    const parsing = {
+      parse: this.parse.bind(this),
+      unparse: this.unparse.bind(this),
+    }, noParsing = { parse: identity, unparse: identity };
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/AbstractDataType.js:87]: Search this.parsing to find out what its for

File: src/editors/_shared/ConfigurationSelector/ConfigurationSelector.jsx [hunk 16]
```diff
@@ -6,48 +6,30 @@ import { SelectScopeModal } from "@mov-ai/mov-fe-lib-react";
...
-  const formatConfigurationValue = configuration => {
-    const document = Document.parsePath(configuration, SCOPES.CONFIGURATION);
-    // Temporary validation if document is from archive
-    // TO BE REMOVED AFTER STANDARDIZATION OF PARSING PROCESS
-    if (document.workspace !== "global") {
-      alert({
-        message: i18n.t("OnlyGlobalConfiguration"),
-        severity: ALERT_SEVERITIES.WARNING
-      });
-    }
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/editors/_shared/ConfigurationSelector/ConfigurationSelector.jsx:38]: Nope.

[the remark this answers is not in the document]
Comment (quirinpa) [reply src/editors/_shared/ConfigurationSelector/ConfigurationSelector.jsx:38]: I had to weigh making human readable vs extremely verbose tests, and I thought that the payoff was worth it.
If you want to understand what you are testing, check the testUtils file. It's not super hard.
Having that gigantic suit of tests verbosely expressed for every data type is a maintainability issue that is greater than that.

File: src/editors/_shared/hooks/DataTypes/types/NumberType.js (hunk 9 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/types/NumberType.js:16]: In other words, the goal is clear, the understanding might not be.

File: src/editors/_shared/hooks/DataTypes/AbstractDataType.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/AbstractDataType.js:92]: textInput means: this input returns text

File: src/editors/_shared/KeyValueTable/ParametersEditorDialog.jsx [hunk 17]
```diff
@@ -119,42 +90,34 @@ const ParameterEditorDialog = props => {
...
       }
-      setData(prevState => {
-        return { ...prevState, value: value };
-      });
+      setData(prevState => ({
+        ...prevState,
+        value: valueToSave({
+          ...prevState,
+          value
+        }),
```
[the remark this answers is not in the document]
Comment (quirinpa) [reply src/editors/_shared/KeyValueTable/ParametersEditorDialog.jsx:106]: Fixed it anyway

File: src/editors/_shared/hooks/DataTypes/testUtils.js (hunk 5 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/testUtils.js:26]: I don't want a complete interaction as I want the mock function to be called only once. Not once for every letter.

File: src/editors/_shared/hooks/DataTypes/types/AnyType.test.js (hunk 6 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/types/AnyType.test.js:14]: We know where things fail much better than if it were inplace, actually, because the tests tell you the index. Believe me, I've tried. We don't have to know by heart, we can look at testUtils

File: src/editors/_shared/hooks/DataTypes/types/ArrayType.js (hunk 7 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/types/ArrayType.js:30]: I don't know what you mean

**Review by quirinpa — COMMENTED** [review]
Fix changelog

File: src/editors/_shared/hooks/DataTypes/AbstractDataType.js [hunk 18]
```diff
@@ -1,160 +1,189 @@
...
+  />);
+}
+
+function CodeEdit(props) {
+  const { isNew, disabled, dataType, ...rest } = useTextEdit(props);
+
+  return (<Typography
+    data-testid="section_data-type-code-editor"
+    component="div"
+    style={{ height: "100px", width: "100%" }}
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/AbstractDataType.js:51]: Can we move this into css?

File: src/editors/_shared/hooks/DataTypes/AbstractDataType.js [hunk 19]
```diff
@@ -1,160 +1,189 @@
...
    */
-  defaultStringEditor(props, mode = "row") {
-    const editor = {
-      row: _props => this.stringEditComponent(_props, ""),
-      dialog: this.codeEditComponent
-    };
-    return editor[mode](props);
+  async validate(value) {
+    // "None" indicates a disabled value
+    if (value === "None")
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/AbstractDataType.js:161]:
```suggestion
    if (value === DISABLED_VALUE)
```
This way you can also remove the comment above.

File: src/editors/_shared/hooks/DataTypes/testUtils.js [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+      expect(type.inputParsing.unparse(parsed)).toStrictEqual(unparsed);
+    }
+  });
+
+  test("Validation (stringOutput)", () => {
+    const type = new DataType({ stringOutput: true });
+
+    validateSuit(type, returns, 26);
+
+    type.validate("None").then(res => expect(res.success).toBe(returns[52]));
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/testUtils.js:98]:
```suggestion
    type.validate(DISABLED_VALUE).then(res => expect(res.success).toBe(returns[52]));
```

File: src/editors/_shared/hooks/DataTypes/testUtils.js (hunk 5 above)
Comment (manuelnogueiramov) [reply src/editors/_shared/hooks/DataTypes/testUtils.js:26]: I still think we need to test with userEvent instead of fireEvent. The more close to reality we get, the more resilient the test.

File: src/editors/_shared/hooks/DataTypes/types/BooleanType.js [hunk 21]
```diff
@@ -1,94 +1,44 @@
...
+  unparse(value) {
+    return value ? "True" : "False";
+  }
 
+  editComponent(props) {
+    const { label, disabled, onChange, rowData = {} } = props;
     return (
       <Checkbox
-        data-testid={"Type=boolean-" + this.label}
+        inputProps={{ "data-testid": "bool-checkbox" }}
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/types/BooleanType.js:33]:
```suggestion
        inputProps={{ "data-testid": "input_bool-checkbox" }}
```

File: src/editors/_shared/hooks/DataTypes/types/BooleanType.js [hunk 22]
```diff
@@ -1,94 +1,44 @@
...
+        data-testid={"Type=boolean-" + (label ?? this.label)}
```
Comment (manuelnogueiramov) [inline src/editors/_shared/hooks/DataTypes/types/BooleanType.js:34]:
```suggestion
        data-testid={"input_Type=boolean-" + (label ?? this.label)}
```

**Review by manuelnogueiramov — DISMISSED** [review]
I left some comments, none of them are blocking. 
I still think these [tests](https://github.com/MOV-AI/frontend-npm-lib-ide/pull/338#discussion_r1777642967) are going to be hard to fix or understand why they are failing if sometime in the future we have to make some changes to any DataType. But I won't hold the PR on this.

This is some good work right here @quirinpa. 
Please still take a look at my comments, as they are easy suggestions. Other than that, have my :heavy_check_mark:

File: src/editors/_shared/hooks/DataTypes/testUtils.js (hunk 20 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/testUtils.js]: Done

File: src/editors/_shared/hooks/DataTypes/types/BooleanType.js (hunk 21 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/types/BooleanType.js]: Done

Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/types/BooleanType.js]: Done

File: src/editors/_shared/hooks/DataTypes/AbstractDataType.js (hunk 18 above)
Comment (quirinpa) [reply src/editors/_shared/hooks/DataTypes/AbstractDataType.js:52]: Yes, we can. It wasn't before, but we can.

**Review by diasdm — APPROVED** [review]

**Review by diasdm — APPROVED** [review]
